File: backend/scripts/create-dynamodb-tables.py
#!/usr/bin/env python3
"""
创建DynamoDB表的脚本
使用AWS凭证环境变量来连接AWS
"""

import json
import boto3
import sys
import os
from botocore.exceptions import ClientError

def load_table_schemas():
    """加载表结构配置"""
    schema_file = os.path.join(os.path.dirname(__file__), '..', 'dynamodb', 'table-schemas.json')
    try:
        with open(schema_file, 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        print(f"错误: 找不到配置文件 {schema_file}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"错误: JSON格式错误 {e}")
        sys.exit(1)

def create_table(dynamodb_client, table_config):
    """创建单个DynamoDB表"""
    table_name = table_config['TableName']
    
    try:
        # 检查表是否已存在
        try:
            dynamodb_client.describe_table(TableName=table_name)
            print(f"✓ 表 {table_name} 已存在")
            return True
        except ClientError as e:
            if e.response['Error']['Code'] != 'ResourceNotFoundException':
                raise
        
        # 创建表
        print(f"📋 正在创建表: {table_name}")
        
        response = dynamodb_client.create_table(**table_config)
        
        # 等待表创建完成
        print(f"⏳ 等待表 {table_name} 创建完成...")
        waiter = dynamodb_client.get_waiter('table_exists')
        waiter.wait(TableName=table_name)
        
        print(f"✅ 表 {table_name} 创建成功")
        return True
        
    except ClientError as e:
        print(f"❌ 创建表 {table_name} 失败: {e.response['Error']['Message']}")
        return False
    except Exception as e:
        print(f"❌ 创建表 {table_name} 时发生未知错误: {str(e)}")
        return False

def main():
    """主函数"""
    print("🚀 开始创建DynamoDB表...")
    
    # 检查AWS凭证环境变量
    required_env_vars = ['AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY', 'AWS_SESSION_TOKEN']
    missing_vars = [var for var in required_env_vars if not os.environ.get(var)]
    
    if missing_vars:
        print(f"❌ 缺少AWS凭证环境变量: {', '.join(missing_vars)}")
        print("请先设置AWS环境变量:")
        print("$env:AWS_ACCESS_KEY_ID=\"your_access_key\"")
        print("$env:AWS_SECRET_ACCESS_KEY=\"your_secret_key\"")
        print("$env:AWS_SESSION_TOKEN=\"your_session_token\"")
        sys.exit(1)
    
    # 创建DynamoDB客户端
    try:
        region = os.environ.get('AWS_DEFAULT_REGION', 'ap-southeast-1')
        dynamodb_client = boto3.client('dynamodb', region_name=region)
        
        # 测试连接
        print(f"🔗 连接AWS DynamoDB (区域: {region})")
        dynamodb_client.list_tables()
        print("✅ AWS连接成功")
        
    except Exception as e:
        print(f"❌ AWS连接失败: {str(e)}")
        print("请检查AWS凭证和网络连接")
        sys.exit(1)
    
    # 加载表配置
    schemas = load_table_schemas()
    tables = schemas.get('tables', [])
    
    if not tables:
        print("❌ 没有找到表配置")
        sys.exit(1)
    
    # 创建所有表
    success_count = 0
    total_count = len(tables)
    
    for table_config in tables:
        if create_table(dynamodb_client, table_config):
            success_count += 1
    
    # 总结结果
    print(f"\n📊 创建结果: {success_count}/{total_count} 个表创建成功")
    
    if success_count == total_count:
        print("🎉 所有DynamoDB表创建完成！")
        
        # 显示创建的表
        print("\n📋 已创建的表:")
        for table_config in tables:
            table_name = table_config['TableName']
            print(f"  • {table_name}")
        
        sys.exit(0)
    else:
        print("⚠️  部分表创建失败，请检查错误信息")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: backend/scripts/create-dynamodb-tables.js
#!/usr/bin/env node
/**
 * 创建DynamoDB表的Node.js脚本
 * 使用AWS凭证环境变量来连接AWS
 */

const fs = require('fs');
const path = require('path');

// 检查环境变量
const requiredEnvVars = ['AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY', 'AWS_SESSION_TOKEN'];
const missingVars = requiredEnvVars.filter(varName => !process.env[varName]);

if (missingVars.length > 0) {
    console.log(`❌ 缺少AWS凭证环境变量: ${missingVars.join(', ')}`);
    console.log('请先设置AWS环境变量:');
    console.log('$env:AWS_ACCESS_KEY_ID="your_access_key"');
    console.log('$env:AWS_SECRET_ACCESS_KEY="your_secret_key"');
    console.log('$env:AWS_SESSION_TOKEN="your_session_token"');
    process.exit(1);
}

// 动态导入AWS SDK
async function createTables() {
    try {
        console.log('🚀 开始创建DynamoDB表...');
        
        // 尝试导入AWS SDK
        let AWS;
        try {
            AWS = require('aws-sdk');
        } catch (error) {
            console.log('❌ AWS SDK未安装，正在安装...');
            console.log('请运行: npm install aws-sdk');
            
            // 创建临时的package.json和安装脚本
            createInstallScript();
            return;
        }

        // 配置AWS
        const region = process.env.AWS_DEFAULT_REGION || 'ap-southeast-1';
        AWS.config.update({ region });

        const dynamodb = new AWS.DynamoDB();

        console.log(`🔗 连接AWS DynamoDB (区域: ${region})`);
        
        // 测试连接
        try {
            await dynamodb.listTables().promise();
            console.log('✅ AWS连接成功');
        } catch (error) {
            console.log(`❌ AWS连接失败: ${error.message}`);
            console.log('请检查AWS凭证和网络连接');
            process.exit(1);
        }

        // 加载表配置
        const schemasPath = path.join(__dirname, '..', 'dynamodb', 'table-schemas.json');
        const schemas = JSON.parse(fs.readFileSync(schemasPath, 'utf8'));
        const tables = schemas.tables || [];

        if (tables.length === 0) {
            console.log('❌ 没有找到表配置');
            process.exit(1);
        }

        // 创建所有表
        let successCount = 0;
        const totalCount = tables.length;

        for (const tableConfig of tables) {
            const success = await createTable(dynamodb, tableConfig);
            if (success) successCount++;
        }

        // 总结结果
        console.log(`\n📊 创建结果: ${successCount}/${totalCount} 个表创建成功`);

        if (successCount === totalCount) {
            console.log('🎉 所有DynamoDB表创建完成！');
            console.log('\n📋 已创建的表:');
            tables.forEach(table => {
                console.log(`  • ${table.TableName}`);
            });
        } else {
            console.log('⚠️  部分表创建失败，请检查错误信息');
            process.exit(1);
        }

    } catch (error) {
        console.log(`❌ 脚本执行失败: ${error.message}`);
        process.exit(1);
    }
}

async function createTable(dynamodb, tableConfig) {
    const tableName = tableConfig.TableName;
    
    try {
        // 检查表是否已存在
        try {
            await dynamodb.describeTable({ TableName: tableName }).promise();
            console.log(`✓ 表 ${tableName} 已存在`);
            return true;
        } catch (error) {
            if (error.code !== 'ResourceNotFoundException') {
                throw error;
            }
        }

        // 创建表
        console.log(`📋 正在创建表: ${tableName}`);
        
        await dynamodb.createTable(tableConfig).promise();
        
        // 等待表创建完成
        console.log(`⏳ 等待表 ${tableName} 创建完成...`);
        await dynamodb.waitFor('tableExists', { TableName: tableName }).promise();
        
        console.log(`✅ 表 ${tableName} 创建成功`);
        return true;
        
    } catch (error) {
        console.log(`❌ 创建表 ${tableName} 失败: ${error.message}`);
        return false;
    }
}

function createInstallScript() {
    const packageJson = {
        "name": "fake-food-detector-backend",
        "version": "1.0.0",
        "description": "Backend scripts for Fake Food Detector",
        "scripts": {
            "create-tables": "node scripts/create-dynamodb-tables.js"
        },
        "dependencies": {
            "aws-sdk": "^2.1691.0"
        }
    };

    const backendPath = path.join(__dirname, '..');
    fs.writeFileSync(
        path.join(backendPath, 'package.json'),
        JSON.stringify(packageJson, null, 2)
    );

    console.log('📦 已创建 backend/package.json');
    console.log('请运行以下命令安装依赖:');
    console.log('cd backend');
    console.log('npm install');
    console.log('然后重新运行此脚本');
}

// 运行脚本
createTables();
